File: Difficulty: Medium/Top View of Binary Tree/top-view-of-binary-tree.cpp
/*
struct Node
{
    int data;
    Node* left;
    Node* right;
};
*/
class Solution {
  public:
    // Function to return a list of nodes visible from the top view
    // from left to right in Binary Tree.
    void dfs(Node* root,int cord,int level, map<int,pair<int,int>>& mp) {
        if(root==nullptr) return;
        if(mp.find(cord)==mp.end()) {
            mp.insert({cord,{level,root->data}});
        }
        else {
            if(mp[cord].first>level) {
                mp[cord].first = level;
                mp[cord].second = root->data;
            }
        }
        dfs(root->left,cord-1,level+1,mp);
        dfs(root->right,cord+1,level+1,mp);
    }
    vector<int> topView(Node *root) {
        
        // code here
        int co_ordinates = 0;
        int level = 0;
        map<int,pair<int,int> > mp;
        dfs(root,co_ordinates,level,mp);
        vector<int> ans;
        for(auto it:mp) {
            ans.push_back(it.second.second);
        }
        return ans;
        
    }
};

File: Difficulty: Medium/Top View of Binary Tree/top-view-of-binary-tree.java
/*
class Node {
    int data;
    Node left, right;

    Node(int val) {
        this.data = val;
        this.left = null;
        this.right = null;
    }
}
*/

class Pair<T,P> {
    public T first;
    public P second;
    public Pair(T f, P s) {
        this.first = f;
        this.second = s;
    }
    
    
}

class Solution {
    private void solve(Node root, int level, int dir, TreeMap<Integer,Pair<Integer,Integer>> ans) {
        if(root==null) return;
        if(ans.containsKey(dir)) {
            if(ans.get(dir).first>level) {
                ans.get(dir).first=level;
                ans.get(dir).second = root.data;
            }
        }
        else {
            ans.put(dir,new Pair<Integer,Integer>(level,root.data));
        }
        solve(root.left,level+1,dir-1,ans);
        solve(root.right,level+1,dir+1,ans);
        
    }
    public ArrayList<Integer> topView(Node root) {
        // code here
        TreeMap<Integer,Pair<Integer,Integer>> ans = new TreeMap<>();
        ArrayList<Integer> ans2 = new ArrayList<>();
        solve(root,0,0,ans);
       // System.out.println(ans);
        for(int a: ans.keySet()) {
            //System.out.print(ans.get(a)+" ");
            ans2.add(ans.get(a).second);
        }
        return ans2;
    }
}